In [0]:
from google.colab import drive
drive.mount("/content/drive")

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
import numpy as np
import pandas as pd

####Reading the bank dataset.

In [0]:
bank_df = pd.read_csv("/content/drive/My Drive/AIML/Residency7th/Lab/ExternalLab2/bank.csv")

In [0]:
bank_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null int64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


####Dropping the columns which are unique for all users like IDs.

In [0]:
bank_df = bank_df.drop(columns = ['RowNumber','CustomerId','Surname'],axis =1)

In [0]:
bank_df.head()

CreditScore Geography  Gender  ...  IsActiveMember  EstimatedSalary  Exited
0          619    France  Female  ...               1        101348.88       1
1          608     Spain  Female  ...               1        112542.58       0
2          502    France  Female  ...               0        113931.57       1
3          699    France  Female  ...               0         93826.63       0
4          850     Spain  Female  ...               1         79084.10       0

[5 rows x 11 columns]

####Distinguishing the feature and target set:

In [0]:
bank_df =  pd.get_dummies(bank_df,drop_first=True)

In [0]:
X = bank_df.drop(columns=['Exited'],axis=1)
y =  bank_df[['Exited']]

####Splitting the data set into Train and test sets:

In [0]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state = 0)

In [0]:
X_train.head()

CreditScore  Age  Tenure  ...  Geography_Germany  Geography_Spain  Gender_Male
7681          641   33       2  ...                  0                0            1
9031          541   39       7  ...                  0                0            1
3691          590   76       5  ...                  0                0            0
202           516   50       5  ...                  0                1            1
5625          508   60       7  ...                  0                0            0

[5 rows x 11 columns]

####Normalizing the train and test data:

In [0]:
#feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()

In [0]:
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

In [0]:
import keras
from keras.models import Sequential
from keras.layers import Dense,Dropout

####Initializing & building the model 

In [0]:
#initialize ANN
classifier = Sequential()


#Adding Input layer and hidden layer
classifier.add(Dense(units= 8, kernel_initializer='uniform',activation='relu', input_dim= 11))
classifier.add(Dropout(0.3))

#Addeding Second hidden layer 
classifier.add(Dense(units=6,kernel_initializer='uniform', activation='relu'))
classifier.add(Dropout(0.5))

#adding output layer
classifier.add(Dense(units=1, kernel_initializer='uniform', activation='sigmoid'))


####Optimizing the model using 'adam' optimizer:

In [0]:
classifier.compile(optimizer='adam',loss='binary_crossentropy', metrics=['accuracy'])

In [0]:
classifier.fit(X_train,y_train, batch_size=30, epochs= 100)

Epoch 1/100
7000/7000 [==============================] - 1s 208us/step - loss: 0.6211 - acc: 0.7961
Epoch 2/100
7000/7000 [==============================] - 0s 50us/step - loss: 0.4773 - acc: 0.7977
Epoch 3/100
7000/7000 [==============================] - 0s 51us/step - loss: 0.4614 - acc: 0.7977
Epoch 4/100
7000/7000 [==============================] - 0s 48us/step - loss: 0.4530 - acc: 0.8063
Epoch 5/100
7000/7000 [==============================] - 0s 48us/step - loss: 0.4511 - acc: 0.8067
Epoch 6/100
7000/7000 [==============================] - 0s 48us/step - loss: 0.4547 - acc: 0.8041
Epoch 7/100
7000/7000 [==============================] - 0s 50us/step - loss: 0.4475 - acc: 0.8064
Epoch 8/100
7000/7000 [==============================] - 0s 50us/step - loss: 0.4529 - acc: 0.8039
Epoch 9/100
7000/7000 [==============================] - 0s 48us/step - loss: 0.4460 - acc: 0.8089
Epoch 10/100
7000/7000 [==============================] - 0s 50us/step - loss: 0.4394 - acc: 0.8119
Epoch 11

####Predicting the results using 0.5 as a threshold: 

In [0]:
y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)

####Preparing the confusion matrix and calculating the Accuracy score:

In [0]:
from sklearn.metrics import confusion_matrix,accuracy_score
cm = confusion_matrix(y_test, y_pred)
acc = accuracy_score(y_test,y_pred)

####The model has provided 81% of accuracy.

In [0]:
acc

0.8113333333333334

In [0]:
cm

array([[2376,    3],
       [ 563,   58]])

####The confusion matrix indicates the following, 
a. The model corectly predicted 2376 employee will leave bank(TP). 

b. 58 employees are idenitified as they will leave the bank, but they won't(TN). 

c. 563 employees are wrongly identified that they will leave the bank(FP).

d. 3 employees are identified as they will not leave the bank, but they will leave the bank(FN)